src/LuxCore.jl

```diff
@@ -105,12 +105,34 @@ Return the input size of the layer.
 """
 function inputsize end
 
+__size(x::AbstractVector{T}) where {T} = isbitstype(T) ? size(x) : __size.(x)
+function __size(x::AbstractArray{T, N}) where {T, N}
+    return isbitstype(T) ? size(x)[1:(N - 1)] : __size.(x)
+end
+__size(x::Tuple) = __size.(x)
+__size(x::NamedTuple{fields}) where {fields} = NamedTuple{fields}(__size.(values(x)))
+__size(x) = fmap(__size, x)
+
 """
-    outputsize(layer)
+    outputsize(layer, x, rng)
 
-Return the output size of the layer.
+Return the output size of the layer. If `outputsize(layer)` is defined, that method
+takes precedence, else we compute the layer output to determine the final size.
 """
-function outputsize end
+function outputsize(layer, x, rng)
+    has_static_outputsize = hasmethod(outputsize, Tuple{typeof(layer)})
+    return outputsize(Val(has_static_outputsize), layer, x, rng)
+end
+
+function outputsize(::Val{true}, layer, x, rng)
+    return outputsize(layer)
+end
+
+function outputsize(::Val{false}, layer, x, rng)
+    ps, st = LuxCore.setup(rng, layer)
+    y = first(apply(layer, x, ps, st))
+    return __size(y)
+end
```

Comment:
```suggestion
function outputsize(layer, x, rng)
    hasmethod(outputsize, Tuple{typeof(layer)}) && return outputsize(layer)
    ps, st = LuxCore.setup(rng, layer)
    y = first(apply(layer, x, ps, st))
    return __size(y)
end
```

Reply:
ah, we don't need to spell this out for the compiler?